helm-browse-project: handle nested projects (Hg inside Git or vice versa)

taquangtrung opened the pull request:
Hi,

This PR improves the logic to choose the most relevant project type in a nested-project scenario, like a Git project residing inside an Hg project, or vice versa.

I modified `helm-browse-project` to choose the project root that is the closet to the current buffer.

thierryvolpiatto commented:
Quang Trung Ta ***@***.***> writes:

> Hi,
>
> This PR improves the logic to choose the most relevant project type in
> a nested-project scenario, like a Git project residing inside an Hg
> project, or vice versa.

I was not aware of such a scenario possible.

> I modified helm-browse-project to choose the project root that is the closet to the current buffer.

Looks good, however could you avoid usage of when-let*? e.g.

Instead of:

    (when-let* ((_ (require 'helm-ls-git nil t))
                (_ (fboundp 'helm-ls-git-root-dir))
                (dir (helm-ls-git-root-dir)))
      (expand-file-name dir))

Use:

    (helm-aif (and (require 'helm-ls-git nil t)
                   (fboundp 'helm-ls-git-root-dir)
                   (helm-ls-git-root-dir))
        (expand-file-name it))

Thanks.

> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/emacs-helm/helm/pull/2723
>
> Commit Summary
>
>   • cbaed45 helm-browse-project: handle nested projects (Hg inside Git or vice versa)
>
> File Changes
>
> (1 file)
>
>   • M helm-files.el (35)
>
> Patch Links:
>
>   • https://github.com/emacs-helm/helm/pull/2723.patch
>   • https://github.com/emacs-helm/helm/pull/2723.diff
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.*Message ID: ***@***.***>

-- 
Thierry

taquangtrung commented:
> Quang Trung Ta ***@***.***> writes:
> Hi, This PR improves the logic to choose the most relevant project type in a nested-project scenario, like a Git project residing inside an Hg project, or vice versa.
> I was not aware of such a scenario possible.
> I modified helm-browse-project to choose the project root that is the closet to the current buffer.
> Looks good, however could you avoid usage of when-let*? e.g. Instead of: (when-let* ((_ (require 'helm-ls-git nil t)) (_ (fboundp 'helm-ls-git-root-dir)) (dir (helm-ls-git-root-dir))) (expand-file-name dir)) Use: (helm-aif (and (require 'helm-ls-git nil t) (fboundp 'helm-ls-git-root-dir) (helm-ls-git-root-dir)) (expand-file-name it)) Thanks.
> […](#)
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ You can view, comment on, or merge this pull request online at: #2723 Commit Summary • [cbaed45](https://github.com/emacs-helm/helm/commit/cbaed45fc1af531f5b190eba62fa8c486678dbcd) helm-browse-project: handle nested projects (Hg inside Git or vice versa) File Changes (1 file) • M helm-files.el (35) Patch Links: • https://github.com/emacs-helm/helm/pull/2723.patch • https://github.com/emacs-helm/helm/pull/2723.diff — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you are subscribed to this thread.*Message ID: ***@***.***>
> -- Thierry

I just updated the patch. Thank you for the tip on `helm-aif`!

Such a scenario happened to me since I put all my dotfiles in a Git repo rooted in my HOME folder.
Then, one of my working projects is located inside HOME using Hg. 

Not a recommended way to store dotfiles, but it's convenient for me :sweat_smile:

thierryvolpiatto commented:
Thanks! Merged.
